Add FutureWarning to all validators

thekaranacharya opened the pull request:
- Add general warning during import (during `from guardrails.validators import <ValidatorName>`)
- Add specific warning during each validator initialization (within each `__init__()` method of the validator)


## Example warnings screenshot
- e.g. User attempting import and usage of `BugFreePython` validator
- Both warnings will be raised:
<img width="1094" alt="Screenshot 2024-03-13 at 2 20 39 PM" src="https://github.com/guardrails-ai/guardrails/assets/27991278/e88c867b-1185-4139-8d4a-0dfba62c85cf">

- Notice how it recommends to use the new naming `ValidPython`, now!

thekaranacharya committed:
Add FutureWarning in __init__.py

thekaranacharya committed:
Add FutureWarnings to all validators

thekaranacharya committed:
Lint fix

thekaranacharya committed:
Add kwargs

CalebCourier commented on guardrails/validators/__init__.py:
```diff
@@ -46,6 +47,19 @@
+    """
+    Importing validators from `guardrails.validators` is deprecated.
+    All validators are now available in Guardrails Hub. Please install
+    and import them from Guardrails Hub instead. All validators would be
```
`would be` -> `will be`

thekaranacharya replied:
Thanks! will vs would is something I always mess up and get confused.

CalebCourier commented on guardrails/validators/__init__.py:
```diff
@@ -46,6 +47,19 @@
+warn(
+    """
+    Importing validators from `guardrails.validators` is deprecated.
+    All validators are now available in Guardrails Hub. Please install
```
`available in Guardrails Hub` -> `available in the Guardrails Hub`

CalebCourier commented on guardrails/validators/__init__.py:
```diff
@@ -46,6 +47,19 @@
+    and import them from Guardrails Hub instead. All validators would be
+    removed from this module in the next major release.
+
+    Install with: `guardrails hub install hub://<namespace>/<validator_name>`
+    Import as: from guardrails.hub import `ValidatorName`
```
I would do a real example here instead of placeholders.
@ShreyaR @zsimjee thoughts?

ShreyaR replied:
@CalebCourier agreed.

Because we have a number of validator name updates I'd suggest making the example specific to the new validator that should be imported from the hub. Instead of the warning here, you can have an f string with each validator's class attributes containing the name and path of the new validator.

thekaranacharya replied:
Got it. The templated f-string can be used in each validator class. The issue with using the same in this file though is, we won't know ahead of time which validator is being imported. Hence this is a general warning here, but I can still add an example.

CalebCourier commented on guardrails/validators/bug_free_python.py:
```diff
@@ -28,6 +29,17 @@ class BugFreePython(Validator):
+            Using this validator from `guardrails.validators` is deprecated.
+            Please install and import this validator from Guardrails Hub instead.
+            This validator would be removed from this module in the next major release.
+            """,
```
If this message will be the same for all of them you can make it a constant so it's easier to search for later.  Probably doesn't matter, but figured I'd throw it out there

ShreyaR replied:
+1. Would also recommend using the same strategy of a templated f-string and the specific validator name here.

thekaranacharya replied:
Got it.

thekaranacharya commented:
Hello @CalebCourier and @ShreyaR ,
In regards to your comments - we can still add a real example and a templated f-string style warning in each validator class as we would know exactly which validator is being used. I see 2 issues with adding an example and templated f-string in `guardrails/validators/__init__.py`:
- We will only know at runtime which validator is being imported. Is there any way we can access that validator name dynamically and use that as an f-string in this file? The warning in `guardrails/validators/__init__.py` will be thrown when any validator is being imported using `from guardrails.validators import <validator_name>`.

- When a validator is imported and used, and if we're using real examples in both places, users will see large **duplicated** warnings. I suggest adding an example only in each validator file (in each __init__ method) - so will be thrown when it's being actively used, and not including an example in `guardrails/validators/__init__.py`. Lmk what you think!

thekaranacharya committed:
Grammar changes

ShreyaR commented:
@thekaranacharya agree that the warning should be raised once. I'd prefer the warning to be raised at import time rather than at initialization, but don't have strong opinions either way.

Here's how you can account for the custom warnings in each validator class:

In the `validator_base` file, you can create a templated deprecated warning like so:

```
validator_import_warning = """Accessing `{validator_name}` through `from guardrails.validators import {validator_name}` is deprecated and support will be removed after version 0.5.x. Please switch to the Guardrails Hub syntax: `from guardrails.hub import {hub_val_name}` for future updates and support. For additional details, visit {hub_val_url}."""
```

Then, in each individual validator, you can specify the variables `validator_name`, `hub_val_name` and `hub_val_url`. If you're raising the warning at import, then these variables can be global variables for the file that contains the specific validator. If you're raising the warning at initialization then these variables can be attributes of the validator class. `validator_name` can probably also just be `class.__name__`.

You can then raise the warning in each validator file like so:

```
warn(
    validator_import_warning.format(validator_name=..., hub_val_name=..., hub_val_url=...),
    ...
)
```

thekaranacharya committed:
Add warning message and mapping from existing OSS validator class names to hub names + hub URLs

thekaranacharya committed:
Add specific warnings in all validators

thekaranacharya committed:
Change get() -> []

thekaranacharya commented:
Hey @ShreyaR and @CalebCourier, this PR has now been updated to reflect this discussion. Added the validator-specific warning with fstrings for each validator which will be thrown during initialization; and added a general warning during import as getting the validator name during import inside `validator/__init__.py` can get nasty looking through import chain.

Please check the updated description.

ShreyaR commented on guardrails/validators/bug_free_python.py:
```diff
@@ -28,6 +31,26 @@ class BugFreePython(Validator):
     | Programmatic fix              | None                              |
     """
 
+    def __init__(self, on_fail: Optional[Callable] = None):
```
@thekaranacharya if you're raising the warning during initialization and the format is common, then you can move this snippet to the parent class of each validator.

During the super().__init__ method the init method of the parent class will be executed and this warning can be raised there instead.

thekaranacharya replied:
Good point! This will actually be a lot cleaner.

thekaranacharya committed:
Refactor warning to validator_base

thekaranacharya committed:
Remove argument, instead get class name directly

thekaranacharya committed:
Remove unnecessary __init__ method

ShreyaR reviewed (approved):
lgtm! :shipit:
